cmd/podman/images/scp.go

```diff
@@ -0,0 +1,383 @@
+			return "", err
+		}
+
+		signers = append(signers, agentSigners...)
+
+		if logrus.IsLevelEnabled(logrus.DebugLevel) {
+			for _, s := range agentSigners {
+				logrus.Debugf("SSH Agent Key %s %s", ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type())
+			}
+		}
+	}
+	var authMethods []ssh.AuthMethod
+	if len(signers) > 0 {
+		var dedup = make(map[string]ssh.Signer)
+		for _, s := range signers {
+			fp := ssh.FingerprintSHA256(s.PublicKey())
+			if _, found := dedup[fp]; found {
+				logrus.Debugf("Dedup SSH Key %s %s", ssh.FingerprintSHA256(s.PublicKey()), s.PublicKey().Type()) // public key info and validation
+			}
+			dedup[fp] = s
+		}
+		var uniq []ssh.Signer
+		for _, s := range dedup {
+			uniq = append(uniq, s)
+		}
+		authMethods = append(authMethods, ssh.PublicKeysCallback(func() ([]ssh.Signer, error) {
+			return uniq, nil
+		}))
+	}
+	if passwdSet { // proper password validation for ssh connection if necessary
+		authMethods = append(authMethods, ssh.Password(passwd))
+	}
+	if len(authMethods) == 0 {
+		authMethods = append(authMethods, ssh.PasswordCallback(func() (string, error) {
+			pass, err := t.ReadPassword(fmt.Sprintf("%s's login password:", uri.User.Username()))
+			return string(pass), err
+		}))
+	}
+	cfg := &ssh.ClientConfig{ // configure ssh client
+		User:            uri.User.Username(),
+		Auth:            authMethods,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	dial, err := ssh.Dial("tcp", uri.Host, cfg) // dial the client
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to connect")
+	}
+	defer dial.Close()
+
+	session, err := dial.NewSession() // connect to the dialed client
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create new ssh session on %q", uri.Host)
+	}
+	defer session.Close()
+	if err != nil {
+		return "", err
+	}
+	if ToFrom { // if we are loading to the remote/ssh client
+		f, _ := os.Open(imageFileIn) // open image file to copy
+		defer f.Close()
+		n, err := scpD.CopyTo(dial, imageFileIn, imageFileOut)
+		if err != nil {
+			fmt.Println(n, "Bytes copied before error. Error while copying file ", err)
+		}
+
+		podman := "podman"
+		if v, found := os.LookupEnv("PODMAN_BINARY"); found {
+			podman = v // get path to podman bin
+		}
+		run := podman + " image load --input=" + imageFileOut // run ssh image load of the file copied via scp
+		var buffer bytes.Buffer
+		session.Stdout = &buffer
+		if err := session.Run(run); err != nil { // run the command declared above
+			return "", err
+		}
+		return buffer.String(), nil // return the output of image load on the machine, should be "Loaded Image(s)..."
+	} // else we are loading from the ssh client
+	podman := "podman"
+	if v, found := os.LookupEnv("PODMAN_BINARY"); found {
+		podman = v // get path to podman bin
+	}
+	run := podman + " image save " + imageFileOut + " --format=oci-archive --output=" + imageFileIn // run ssh image load of the file copied via scp. Files are reverse in thie case...
```

Comment:
Does this handle special characters (including `' '` and what not) in the parameters?

Reply:
Still outstanding I think.

Reply:
You can't specify a file anymore so a blank won't happen I don't think?

Reply:
`podman image scp remote1::'fedora; cat /etc/passwd; echo "pwned" | mail root' remote1::'fedora; touch /pwned/too'`

Sure no-one _sane_ would do this in isolation, but combine it with a web UI that lets some low-privileged external user define an image to run (or rather not explicitly an image, but choose a language from a web form drop-down, but allowing the attacker to submit different values using raw HTTP), and … 

Security practitioners describe such situations as “composition kills”. It’s better not to introduce such risks.

Isn’t there a validation function somewhere in Podman? `c/image/docker/registry.ParseNormalizedNamed` would do as a last resort, but it would also reject the transport:… syntax.